components/engine/client/hijack.go

```diff
@@ -177,12 +177,14 @@ func (cli *Client) setupHijackConn(req *http.Request, proto string) (net.Conn, e
 
 	// Server hijacks the connection, error 'connection closed' expected
 	resp, err := clientconn.Do(req)
-	if err != nil {
-		return nil, err
-	}
-	if resp.StatusCode != http.StatusSwitchingProtocols {
-		resp.Body.Close()
-		return nil, fmt.Errorf("unable to upgrade to %s, received %d", proto, resp.StatusCode)
+	if err != httputil.ErrPersistEOF {
```

Comment:
Perhaps this should be separate from the other changes; this is a back port of https://github.com/moby/moby/pull/34275/commits/7a53991bd6200b875e492d9c3f8e99a9a0c11056 (https://github.com/moby/moby/pull/34275) - "Fix panic in hijack"